google/cloud/internal/credentials.h

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_CREDENTIALS_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_CREDENTIALS_H
+
+#include "google/cloud/version.h"
+#include "absl/types/optional.h"
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+class CredentialsVisitor;
+}  // namespace internal
+
+// TODO(#6293) - move unified credentials out of internal namespace
+namespace internal {
+class Credentials {
+ public:
+  virtual ~Credentials() = 0;
+
+ private:
+  friend class CredentialsVisitor;
+  virtual void dispatch(CredentialsVisitor& visitor) = 0;
+};
+
+std::shared_ptr<Credentials> GoogleDefaultCredentials();
+std::shared_ptr<Credentials> AccessTokenCredentials(
+    std::string const& access_token,
+    std::chrono::system_clock::time_point expiration);
```

Comment:
Is there a reason this argument isn't a `struct AccessToken` (defined below)?

Reply:
I am seriously considering keeping `struct AccessToken` inside the `internal` namespace.